File: examples/python/02_contains.py
import json

import oc_sidememory

from _transformers import accepts, load_runtime, masked_next_token


model, tokenizer, vocabulary = load_runtime()
schema = json.dumps({
    "type": "array",
    "items": {"type": "integer", "enum": [1, 2]},
    "minItems": 2,
    "maxItems": 2,
    "contains": {"const": 1},
    "minContains": 1,
    "maxContains": 1,
})
guide = oc_sidememory.SidememoryGuide(
    oc_sidememory.compile_schema(schema, vocabulary, tokenizer.vocab_size)
)
assert accepts(guide, tokenizer, "[1,2]")
assert not accepts(guide, tokenizer, "[2,2]")
selected = masked_next_token(model, tokenizer, guide)
print(f"contains: real-model token {selected!r}; missing match rejected")

File: examples/rust/02_contains.rs
mod common;

use oc_sidememory::sidememory::ProbeDecision;

fn main() {
    let schema = r#"{"type":"array","items":{"type":"integer","enum":[1,2]},"minItems":2,"maxItems":2,"contains":{"const":1},"minContains":1,"maxContains":1}"#;
    let compiled = common::compile(schema, &[b"[1,2]", b"[2,2]"], None);
    let mut guide = common::guide(compiled);
    assert_eq!(guide.probe(0).expect("probe succeeds"), ProbeDecision::Allow);
    assert!(matches!(guide.probe(1).expect("probe succeeds"), ProbeDecision::Deny(_)));
    println!("contains: accepted one match and rejected zero matches");
}
